File: xauusd_ha_backtest/backtest.py
import requests
import pandas as pd
import time
from datetime import datetime, timedelta
import sys

# We use Binance PAXGUSDT (PAX Gold) as a proxy for spot Gold. It perfectly tracks 1oz Gold and allows downloading exact 3m intervals for 30 days.
symbol = "PAXGUSDT"
interval = "3m"
limit = 1000

end_time = int(time.time() * 1000)
start_time = int((datetime.now() - timedelta(days=30)).timestamp() * 1000)

all_klines = []
current_start = start_time

print("Downloading 30 days of 3m Gold data...")
while current_start < end_time:
    url = f"https://api.binance.com/api/v3/klines?symbol={symbol}&interval={interval}&startTime={current_start}&limit={limit}"
    try:
        res = requests.get(url)
        data = res.json()
        if not data:
            break
        all_klines.extend(data)
        current_start = data[-1][0] + 1
        time.sleep(0.1)
    except Exception as e:
        print("Error fetching data:", e)
        sys.exit(1)

df = pd.DataFrame(all_klines, columns=['open_time', 'open', 'high', 'low', 'close', 'volume', 'close_time', 'qav', 'num_trades', 'taker_base_vol', 'taker_quote_vol', 'ignore'])
df['open_time'] = pd.to_datetime(df['open_time'], unit='ms')
df = df[['open_time', 'open', 'high', 'low', 'close']].astype({'open': float, 'high': float, 'low': float, 'close': float})
df.to_csv('xauusd_3m_30days.csv', index=False)
print(f"Downloaded {len(df)} candles.")

# Calculate HA
ha_close = (df['open'] + df['high'] + df['low'] + df['close']) / 4
ha_open = [ (df['open'].iloc[0] + df['close'].iloc[0]) / 2 ]

for i in range(1, len(df)):
    ha_open.append( (ha_open[i-1] + ha_close.iloc[i-1]) / 2 )

df['ha_close'] = ha_close
df['ha_open'] = ha_open
df['ha_high'] = df[['high', 'ha_open', 'ha_close']].max(axis=1)
df['ha_low'] = df[['low', 'ha_open', 'ha_close']].min(axis=1)

df['ha_color'] = df.apply(lambda row: 'Green' if row['ha_close'] > row['ha_open'] else 'Red', axis=1)

# Strategy rules
MAX_CANDLE_SIZE = 2.5 # Assuming 25 points = 250 pips = $2.5 move in Gold

position = None
entry_price = 0.0
entry_time = None
trades = []

for i in range(1, len(df)):
    prev_color = df['ha_color'].iloc[i-1]
    curr_color = df['ha_color'].iloc[i]

    curr_high = df['high'].iloc[i]
    curr_low = df['low'].iloc[i]
    curr_close = df['close'].iloc[i]
    curr_time = df['open_time'].iloc[i]

    candle_size = curr_high - curr_low

    buy_signal = (prev_color == 'Red') and (curr_color == 'Green')
    sell_signal = (prev_color == 'Green') and (curr_color == 'Red')

    is_oversized = candle_size > MAX_CANDLE_SIZE

    # Check Exits
    if position == 'Long' and sell_signal:
        pnl = curr_close - entry_price
        trades.append({'Type': 'Long', 'Entry_Time': str(entry_time), 'Entry_Price': entry_price, 'Exit_Time': str(curr_time), 'Exit_Price': curr_close, 'PnL': pnl, 'Oversized_Exit': is_oversized})
        position = None

    elif position == 'Short' and buy_signal:
        pnl = entry_price - curr_close
        trades.append({'Type': 'Short', 'Entry_Time': str(entry_time), 'Entry_Price': entry_price, 'Exit_Time': str(curr_time), 'Exit_Price': curr_close, 'PnL': pnl, 'Oversized_Exit': is_oversized})
        position = None

    # Check Entries
    if position is None:
        if buy_signal and not is_oversized:
            position = 'Long'
            entry_price = curr_close
            entry_time = curr_time
        elif sell_signal and not is_oversized:
            position = 'Short'
            entry_price = curr_close
            entry_time = curr_time

# Summary
total_trades = len(trades)
winning_trades = len([t for t in trades if t['PnL'] > 0])
losing_trades = len([t for t in trades if t['PnL'] <= 0])
win_rate = (winning_trades / total_trades * 100) if total_trades > 0 else 0
total_pnl = sum([t['PnL'] for t in trades])

with open('backtest_results.md', 'w') as f:
    f.write("# XAUUSD 3-Minute Heikin Ashi Strategy Backtest\n\n")
    f.write(f"**Data Period**: Last 30 Days (3-minute candles)\n")
    f.write(f"**Filter applied**: Candle size > 25 points ($2.5) -> Exit but no entry.\n\n")
    f.write("## Performance Summary\n")
    f.write(f"- **Total Trades**: {total_trades}\n")
    f.write(f"- **Winning Trades**: {winning_trades}\n")
    f.write(f"- **Losing Trades**: {losing_trades}\n")
    f.write(f"- **Win Rate**: {win_rate:.2f}%\n")
    f.write(f"- **Net PnL (in USD terms per 1 oz)**: ${total_pnl:.2f}\n\n")
    f.write("## Sample of Last 10 Trades\n")
    f.write("| Type | Entry Time | Entry Price | Exit Time | Exit Price | PnL | Oversized Exit |\n")
    f.write("|---|---|---|---|---|---|---|\n")
    for t in trades[-10:]:
        f.write(f"| {t['Type']} | {t['Entry_Time']} | {t['Entry_Price']:.2f} | {t['Exit_Time']} | {t['Exit_Price']:.2f} | {t['PnL']:.2f} | {t['Oversized_Exit']} |\n")
print("Backtest complete. Results saved to backtest_results.md")

File: xauusd_ha_backtest/backtest.js
const fs = require('fs');
const https = require('https');

function getBinanceData(symbol, interval, startTime, limit) {
    return new Promise((resolve, reject) => {
        const url = `https://api.binance.com/api/v3/klines?symbol=${symbol}&interval=${interval}&startTime=${startTime}&limit=${limit}`;
        https.get(url, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => resolve(JSON.parse(data)));
        }).on('error', reject);
    });
}

async function run() {
    const symbol = "PAXGUSDT"; // Proxy for Gold since it tracks 1oz Gold 1:1 and has exact 3m intervals
    const interval = "3m";
    const limit = 1000;
    
    const end_time = Date.now();
    let current_start = Date.now() - (30 * 24 * 60 * 60 * 1000);
    
    let all_klines = [];
    console.log("Downloading 30 days of 3m Gold data...");
    
    while (current_start < end_time) {
        let data = await getBinanceData(symbol, interval, current_start, limit);
        if (!data || data.length === 0) break;
        all_klines.push(...data);
        current_start = data[data.length - 1][0] + 1;
        await new Promise(r => setTimeout(r, 100)); // Sleep
    }
    console.log(`Downloaded ${all_klines.length} candles.`);
    
    // Create DataFrame equivalent
    let df = all_klines.map(k => ({
        open_time: new Date(k[0]).toISOString().replace('T', ' ').substring(0, 19),
        open: parseFloat(k[1]),
        high: parseFloat(k[2]),
        low: parseFloat(k[3]),
        close: parseFloat(k[4])
    }));
    
    // HA Calculation
    let ha_close = df.map(row => (row.open + row.high + row.low + row.close) / 4);
    let ha_open = [(df[0].open + df[0].close) / 2];
    
    for (let i = 1; i < df.length; i++) {
        ha_open.push((ha_open[i-1] + ha_close[i-1]) / 2);
    }
    
    for (let i = 0; i < df.length; i++) {
        df[i].ha_close = ha_close[i];
        df[i].ha_open = ha_open[i];
        df[i].ha_color = df[i].ha_close > df[i].ha_open ? 'Green' : 'Red';
    }
    
    // Save to CSV
    let csvContent = "open_time,open,high,low,close\n" + df.map(row => `${row.open_time},${row.open},${row.high},${row.low},${row.close}`).join("\n");
    fs.writeFileSync('xauusd_3m_30days.csv', csvContent);
    
    const MAX_CANDLE_SIZE = 25.0; // 25 points = $25.0
    
    let position = null;
    let entry_price = 0.0;
    let entry_time = null;
    let trades = [];
    
    for (let i = 1; i < df.length; i++) {
        let prev_color = df[i-1].ha_color;
        let curr_color = df[i].ha_color;
        
        let curr_high = df[i].high;
        let curr_low = df[i].low;
        let curr_close = df[i].close;
        let curr_time = df[i].open_time;
        
        let candle_size = curr_high - curr_low;
        
        let buy_signal = (prev_color === 'Red') && (curr_color === 'Green');
        let sell_signal = (prev_color === 'Green') && (curr_color === 'Red');
        let is_oversized = candle_size > MAX_CANDLE_SIZE;
        
        if (position === 'Long' && sell_signal) {
            let pnl = curr_close - entry_price;
            trades.push({Type: 'Long', Entry_Time: entry_time, Entry_Price: entry_price, Exit_Time: curr_time, Exit_Price: curr_close, PnL: pnl, Oversized_Exit: is_oversized});
            position = null;
        } else if (position === 'Short' && buy_signal) {
            let pnl = entry_price - curr_close;
            trades.push({Type: 'Short', Entry_Time: entry_time, Entry_Price: entry_price, Exit_Time: curr_time, Exit_Price: curr_close, PnL: pnl, Oversized_Exit: is_oversized});
            position = null;
        }
        
        if (position === null) {
            if (buy_signal && !is_oversized) {
                position = 'Long';
                entry_price = curr_close;
                entry_time = curr_time;
            } else if (sell_signal && !is_oversized) {
                position = 'Short';
                entry_price = curr_close;
                entry_time = curr_time;
            }
        }
    }
    
    let winning_trades = trades.filter(t => t.PnL > 0).length;
    let losing_trades = trades.filter(t => t.PnL <= 0).length;
    let win_rate = trades.length > 0 ? (winning_trades / trades.length * 100) : 0;
    let total_pnl = trades.reduce((sum, t) => sum + t.PnL, 0);
    
    let md = `# XAUUSD 3-Minute Heikin Ashi Strategy Backtest\n\n`;
    md += `**Data Period**: Last 30 Days (3-minute candles)\n`;
    md += `**Filter applied**: Candle size > 25 points ($25.0) -> Exit but no entry.\n\n`;
    md += `## Performance Summary\n`;
    md += `- **Total Trades**: ${trades.length}\n`;
    md += `- **Winning Trades**: ${winning_trades}\n`;
    md += `- **Losing Trades**: ${losing_trades}\n`;
    md += `- **Win Rate**: ${win_rate.toFixed(2)}%\n`;
    md += `- **Net PnL (in USD terms per 1 oz)**: $${total_pnl.toFixed(2)}\n\n`;
    
    md += `## Sample of Last 10 Trades\n`;
    md += `| Type | Entry Time | Entry Price | Exit Time | Exit Price | PnL | Oversized Exit |\n`;
    md += `|---|---|---|---|---|---|---|\n`;
    trades.slice(-10).forEach(t => {
        md += `| ${t.Type} | ${t.Entry_Time} | ${t.Entry_Price.toFixed(2)} | ${t.Exit_Time} | ${t.Exit_Price.toFixed(2)} | ${t.PnL.toFixed(2)} | ${t.Oversized_Exit} |\n`;
    });
    
    fs.writeFileSync('backtest_results.md', md);
    console.log("Backtest complete. Results saved to backtest_results.md");
}

run();
